docs/source/tutorials/qmd/07-groups.qmd

```diff
@@ -0,0 +1,330 @@
+`SplineCoef`:
+
+```{python}
+second_diff = np.diff(np.eye(10), n=2, axis=0)
+penalty = second_diff.T @ second_diff
+coef_group = SplineCoef(name="coef", penalty=penalty, tau2=tau2_group["tau2"])
+```
+
+Each group offers access to all its members through `mappingproxy` instances, which are
+basically immutable dictionaries:
+
+```{python}
+tau2_group.vars
+```
+
+```{python}
+coef_group.vars
+```
+
+## A complete P-Spline group
+
+In this last step, we create overarching group that incorporates the observed
+basis matrix of our spline and uses the parameter groups to evaluate the smooth
+function. Note the following:
+
+1. Here, we create a `SplineCoef` instance *inside* the new group's `__init__`, because
+   this step is always the same with P-Splines, once we know the penalty and the variance
+   parameter $\tau^2$.
+2. We do not automatically create an instance of our `IGVariance` group for the variance
+   parameter, but allow users to pass any fitting group instance to the class. This makes
+   our `PSpline` class very flexible: we could easily replace the `IGVariance` group
+   with another group that represents a different prior for $\tau^2$. For this `PSpline`
+   group, it is only important that there is a key `"tau2"` in the `tau2_group`.
+3. In our `super().__init__` call at the end, we make sure to add not only the
+   Vars `basis_matrix` and `smooth` that we directly created, but also all
+   members of the `coef_group` and `tau2_group`. That way, we can access all
+   relevant vars just from this overarching `PSpline` group.
+
+```{python}
+class PSpline(lsl.Group):
+    def __init__(
+        self, name, basis_matrix: Array, penalty: Array, tau2_group: lsl.Group
+    ) -> None:
+        coef_group = SplineCoef(
+            name=f"{name}_coef", penalty=penalty, tau2=tau2_group["tau2"]
+        )
+
+        basis_matrix = lsl.Obs(basis_matrix, name="basis_matrix")
+        smooth = lsl.Var(
+            lsl.Calc(jnp.dot, basis_matrix, coef_group["coef"]), name=name
+        )
+
+        group_vars = coef_group.nodes_and_vars | tau2_group.nodes_and_vars
+
+        super().__init__(
+            name=name,
+            basis_matrix=basis_matrix,
+            smooth=smooth,
+            **group_vars
+        )
+```
+
+Of course it may be convenient to specialize this `PSpline` group, depending on
+your usage. For example, we could imagine to adapt it to an `IGPSpline` group
+that works on the assumption that you want to use an inverse gamma prior on
+$\tau^2$. Such a group would require the input parameters `a` and `b` instead of
+the `tau2_group`. The latter would be created automatically, just as the
+`coef_group` is in the code snippet above.
+
+### Group graph
+
+To see what we have created, let's initialize a toy example of our group to look
+at the resulting graph. First, we create the group:
+
+```{python}
+basis_matrix = np.random.uniform(size=(15, 10)) # NOT a valid basis matrix, just for show!
```

Comment:
We should change this eventually to a real design_matrix.

Reply:
I agree. I intended to use `scipy`s spline functionality, but I did not find a nice helper function for creating the knots - and pasting a function for that into the tutorial seemed a bit much to me. I think if/once we have the spline functionality form liesel-internal in liesel-core, we should certainly replace this here. Then we can also complete the example in the tutorial by fitting a real model.